feat: Editorial review: Add content to cover the field-sizing property (#32489)
mdn/content: files/en-us/web/css/field-sizing/index.md

---
title: field-sizing
slug: Web/CSS/field-sizing
page-type: css-property
browser-compat: css.properties.field-sizing
---

{{CSSRef}}{{seecompattable}}

The **`field-sizing`** [CSS](/en-US/docs/Web/CSS) property sets the sizing behavior of elements that by default are given a default preferred size — i.e. form control elements. `field-sizing` enables this behavior to be overridden so that controls are instead sized to fit their contents.

This property is typically used to create text [inputs](/en-US/docs/Web/HTML/Element/input) and {{htmlelement("textarea")}}s that shrinkwrap their content and grow as more text is entered.

## Syntax

```css
/* Keyword values */
field-sizing: content;
field-sizing: fixed;

/* Global values */

> empty-cells: inherit;

Reviewer: Need to update the property name.

Author: D'oh, yes! Fixed ;-)

empty-cells: initial;
empty-cells: revert;
empty-cells: revert-layer;
empty-cells: unset;
```

### Values

- `content`
  - : Affected elements are sized to fit their contents.
- `fixed`
  - : Affected elements are sized with a fixed preferred size.

## Description

`field-sizing: content` overrides the default preferred sizing of form elements, providing an easy way to create text inputs that shrinkwrap their content and grow as more text is entered, only stopping their growth and requiring scrolling to view all the content when maximum size limits are reached (i.e. their containing elements' size, or a size set on them via CSS).

Specifically, this property affects the following elements:

> - Form input types that accept direct text input from users: This includes [`email`](/en-US/docs/Web/HTML/Element/input/email), [`number`](/en-US/docs/Web/HTML/Element/input/number), [`password`](/en-US/docs/Web/HTML/Element/input/password), [`search`](/en-US/docs/Web/HTML/Element/input/search), [`tel`](/en-US/docs/Web/HTML/Element/input/tel), [`text`](/en-US/docs/Web/HTML/Element/input/text), and [`url`](/en-US/docs/Web/HTML/Element/input/url) types.

Reviewer: We should mention `size` attribute is ignored if `field-sizing:content`.

Author: Good call. I've added a sentence to cover this.

- {{htmlelement("textarea")}} controls: It is worth noting that `<textarea>` elements with `field-sizing: content` set behave much like single-line text controls do, with the following additions:
  - If `<textarea>`s are unable to grow due to a width constraint, they will start to grow in height to display additional rows of content. When a height constraint is then reached, they will then start to show a scrollbar to allow all the content to be viewed.
  - [`rows`](/en-US/docs/Web/HTML/Element/textarea#cols) and [`cols`](/en-US/docs/Web/HTML/Element/textarea#cols) attributes modify the default preferred size of a `<textarea>`, so as such do not have any effect on `<textarea>` elements with `field-sizing: content` set.
- {{htmlelement("select")}} controls: These behave a bit differently to what you might expect with `field-sizing: content` set. The effect depends on the type of select control you are creating:
  - Regular drop-down select boxes will change width to always fit the displayed option value, as new values are selected. (By default, the drop-down would be sized large enough to display the longest option value.)

>   - List boxes (`<select>`s with the [`multiple`](/en-US/docs/Web/HTML/Element/select#multiple) attribute set) will be large enough to display all the options without needing to scroll. (By default, the drop-down would require scrolling to view all the option values.)

Reviewer: We had better mention `size` attribute is almost ignored.  Browsers checks if the value is `1` or not to switch list box mode and drop-down mode.

Author: Agreed. I have added a new sub-bullet to cover it:

- The [`size`](/en-US/docs/Web/HTML/Element/select#size) attribute has very little effect on `<select>` elements that also have `field-sizing: content` set on them. In such cases, the browser will check whether `size` is equal to `1` or not to determine whether the select should be a drop-down or a listbox, but it will always display all the options of a listbox even if `size` is smaller than the number of options.

---
title: field-sizing
slug: Web/CSS/field-sizing
page-type: css-property
browser-compat: css.properties.field-sizing
---

{{CSSRef}}{{seecompattable}}

> The **`field-sizing`** [CSS](/en-US/docs/Web/CSS) property sets the sizing behavior of elements that by default are given a default preferred size — i.e. form control elements. `field-sizing` enables this behavior to be overridden so that controls are instead sized to fit their contents.

Author: This was largely good, so I used it, but I adjusted it a bit to:

> The **`field-sizing`** [CSS](/en-US/docs/Web/CSS) property enables you to control the sizing behavior of elements that are given a default preferred size, such as form control elements. This property enables you to override the default sizing behavior, allowing form controls to adjust in size to fit their contents.

I didn't like "typically" because:

1. It was used in the next paragraph too.
2. It implies "usually, but not always", whereas by default, form conttrols are always given a default preferred size.

> The **`field-sizing`** [CSS](/en-US/docs/Web/CSS) property sets the sizing behavior of elements that by default are given a default preferred size — i.e. form control elements. `field-sizing` enables this behavior to be overridden so that controls are instead sized to fit their contents.

Reviewer 2: Nice, the final output is good

This property is typically used to style text {{htmlelement("input")}} and {{htmlelement("textarea")}} elements to allow them to shrinkwrap their content as well as grow when more text is entered into the form control.

## Syntax

```css
/* Keyword values */
field-sizing: content;
field-sizing: fixed;

/* Global values */
field-sizing: inherit;
field-sizing: initial;
field-sizing: revert;
field-sizing: revert-layer;
field-sizing: unset;
```

### Values

- `content`
  - : Allows the element to adjust its size to fit its contents.
- `fixed`

>   - : Sets a fixed size for the element.

Author: Good call, however, I've just added "This is the default value." If you look at the computed values on any element, a default `field-sizing: fixed` will be shown.

## Description

> `field-sizing: content` overrides the default preferred sizing of form elements, providing an easy way to create text inputs that shrinkwrap their content and grow as more text is entered, only stopping their growth and requiring scrolling to view all the content when maximum size limits are reached (i.e. their containing elements' size, or a size set on them via CSS).

Author: Mostly good. I fiddled with it a bit and ended up with:

> `field-sizing: content` overrides the default preferred sizing of form elements. This setting provides an easy way to configure text inputs to shrinkwrap their content and grow as more text is entered. They stop expanding when they reach maximum size limits (defined by the size of their containing element or set via CSS), at which point scrolling is required to view all the content.

1. I didn't want to say "property" because a property plus a value is really a declaration; I did like "setting" as a way of describing what you are doing.
2. The parenthetical bit got rather wordy and over-complex, so I simplified it.

> Specifically, this property affects the following elements:

Author: I've altered a couple of bullets in this section with what you said in mind, but I've left most of the instances in, as I think they work OK with how this lead-in sentence is written.

> Specifically, this property affects the following elements:

Reviewer 2: Thanks!

> ### Elements affected by `field-sizing`

Reviewer 3: ```suggestion
### Elements affected by `field-sizing: content`
```

> Specifically, this property affects the following elements:

Reviewer 3: ```suggestion
Specifically, setting this property to `content` affects the following elements:
```

- Form input types that accept direct text input from users. This includes [`email`](/en-US/docs/Web/HTML/Element/input/email), [`number`](/en-US/docs/Web/HTML/Element/input/number), [`password`](/en-US/docs/Web/HTML/Element/input/password), [`search`](/en-US/docs/Web/HTML/Element/input/search), [`tel`](/en-US/docs/Web/HTML/Element/input/tel), [`text`](/en-US/docs/Web/HTML/Element/input/text), and [`url`](/en-US/docs/Web/HTML/Element/input/url) types.

>   - If no minimum width is set on the control, it will only be as wide as the text cursor.

Reviewer 3: ~~need to clarify that this is if `content` is set. Not true if `fixed` is set, where the size is set by the `size` attribute or the browser's default preferred size.~~ If we change the h3 heading to "Elements affected by `field-sizing: content` and intro statement this edit is not needed, and the multiple "elements with `field-sizing: content` set" can be removed from bullets below.

Author: I've made the heading change, however, I think most of those instances are still needed — the sentences read a bit strangely when removed, and some of them are important so that the sentence makes sense.

> - [`file`](/en-US/docs/Web/HTML/Element/input/file) inputs: Direct text input is not possible, but the filename display changes as the user selects a new file to upload, which will change the control size if `field-sizing: content` is set.

Author: I didn't feel like this was quite accurate, so I changed it to this:

> [`file`](/en-US/docs/Web/HTML/Element/input/file) inputs. Direct text input is not possible; however, the displayed filename changes as the user selects a new file to upload. If `field-sizing: content` is set, the control will change size to shrinkwap the filename.

Does that work for you?

> - [`file`](/en-US/docs/Web/HTML/Element/input/file) inputs: Direct text input is not possible, but the filename display changes as the user selects a new file to upload, which will change the control size if `field-sizing: content` is set.

Reviewer 2: It does, yes, this is good! Did we get rid of the colons? I think they were good to have

- {{htmlelement("textarea")}} controls. It is worth noting that `<textarea>` elements with `field-sizing: content` set behave much like single-line text controls, with the following additions:
  - If `<textarea>` elements are unable to grow due to a width constraint, they will start to grow in height to display additional rows of content. When a height constraint is then reached, they will then start to show a scrollbar to allow all the content to be viewed.
  - [`rows`](/en-US/docs/Web/HTML/Element/textarea#cols) and [`cols`](/en-US/docs/Web/HTML/Element/textarea#cols) attributes modify the default preferred size of a `<textarea>`. As a result, `rows`/`cols` have no effect on `<textarea>` elements with `field-sizing: content` set.
- {{htmlelement("select")}} controls. These behave a bit differently to what you might expect with `field-sizing: content` set. The effect depends on the type of `<select>` control you are creating:

>   - Regular drop-down boxes will change their width to always fit the displayed option value as new values are selected. (By default, the dropdown's size is set to be large enough to display the longest option value.)
>   - List boxes (`<select>` elements with the [`multiple`](/en-US/docs/Web/HTML/Element/select#multiple) or [`size`](/en-US/docs/Web/HTML/Element/select#multiple) attributes set) will be large enough to display all the options without needing to scroll. (By default, the drop-down will require scrolling to view all the option values.)

Reviewer 2: Missed these
```suggestion
  - Regular drop-down boxes will change their width to always fit the displayed option value as new values are selected. (By default, the drop-down's size is set to be large enough to display the longest option value.)
  - List boxes (`<select>` elements with the [`multiple`](/en-US/docs/Web/HTML/Element/select#multiple) or [`size`](/en-US/docs/Web/HTML/Element/select#multiple) attribute) will be large enough to display all the options without needing to scroll. (By default, the drop-down box will require scrolling to view all the option values.)
```

>   - Regular drop-down boxes will change their width to always fit the displayed option value as new values are selected. (By default, the dropdown's size is set to be large enough to display the longest option value.)
>   - List boxes (`<select>` elements with the [`multiple`](/en-US/docs/Web/HTML/Element/select#multiple) or [`size`](/en-US/docs/Web/HTML/Element/select#multiple) attributes set) will be large enough to display all the options without needing to scroll. (By default, the drop-down will require scrolling to view all the option values.)

Reviewer 2: Missed these in previous round
```suggestion
  - Regular drop-down boxes will change their width to always fit the displayed option value as new values are selected. (By default, the drop-down's size is set to be large enough to display the longest option value.)
  - List boxes (`<select>` elements with the [`multiple`](/en-US/docs/Web/HTML/Element/select#multiple) or [`size`](/en-US/docs/Web/HTML/Element/select#multiple) attribute) will be large enough to display all the options without needing to scroll. (By default, the drop-down box will require scrolling to view all the option values.)
```

> ### Sizing controls with CSS

Author: Yup, good call. I've given that section its own heading of "### Elements affected by `field-sizing`"

> ### Sizing controls with CSS

Author: Yes! Another issue is that this section also talks about `maxlength`, which is an HTML attribute.

I thought about this for quite a while, and eventually came up with "### `field-sizing` interaction with other size settings"

What do you think?

> ### Sizing controls with CSS

Reviewer 2: It's pretty good and goes very well with the previous H3.
I was apparently restricting myself to the gerund verbiage for the title but we need not because this is not the Examples section.

The sizing imposed on form controls by `field-sizing: content` can be overidden by CSS sizing properties. It doesn't really make sense to set {{cssxref("width")}} and {{cssxref("height")}} because that will reimpose a fixed size on the control, but using properties like {{cssxref("min-width")}} and {{cssxref("max-width")}} work well alongside `field-sizing: content`, allowing the control to grow and shrink with the entered text without getting too big or small.

> It is also useful to combine these properties with form attributes like [`maxlength`](/en-US/docs/Web/HTML/Element/input#maxlength), which will cause the control to stop growing in size when the maximum character limit is reached.

Author: I think what I really mean is "It is also useful to combine `field-sizing: content` with...", so I've changed it to that.

> It is also useful to combine `field-sizing: content` with form attributes like [`maxlength`](/en-US/docs/Web/HTML/Element/input#maxlength), which will cause the control to stop growing in size when the maximum character limit is reached.

Reviewer 3: In most cases, setting maxlength is generally not recommended. A main issue was (and possibly still is) that as the user can keep entering characters, these characters are read by the screen reader, but there is no indication that they aren't being added to the input. There were a lot of discussions about this 5 and 10 years ago. Letting the user add more than the max characters but indicating that they're beyond the limit, color, and not allowing submission (i.e. Twitter, Mastodon, etc. UI pattern) are better usability. Ex.  https://heymichellemac.com/design-pattern-character-limits. So maybe omit this or downplay it because we don't want to be recommending it.

Author: I've downplayed it a bit — removed some of the instances from the examples, and removed language that says it is useful.

{{cssinfo}}

## Formal syntax

{{csssyntax}}

## Examples

### Growing/shrinking text fields

This example illustrates the effect of `field-sizing: content` on single- and multi-line text fields — they change size as text is added and removed, in effect shrinkwrapping the contents, until an lower or upper size limit is reached.

#### HTML

> Our HTML contains three form fields with associated {{htmlelement("label")}}s — `<input>`s of type [`text`](/en-US/docs/Web/HTML/Element/input/text) and [`email`](/en-US/docs/Web/HTML/Element/input/email) respectively, and a {{htmlelement("textarea")}}.

Author: I mostly used this, but I just said "HTML" without the "code". I have accepted a few of your suggestions with "code" because I think it is ok to add variance and make the language a bit more colourful, but at the same time it is pretty obvious it is code to the target audience, so it feels a bit redundant to use it all the time.

> Our HTML contains three form fields with associated {{htmlelement("label")}}s — `<input>`s of type [`text`](/en-US/docs/Web/HTML/Element/input/text) and [`email`](/en-US/docs/Web/HTML/Element/input/email) respectively, and a {{htmlelement("textarea")}}.

Reviewer 2: Makes sense!

```html
<div>
  <label for="text">Enter name:</label>
  <input type="text" id="text" maxlength="50" />
</div>
<div>
  <label for="email">Enter email:</label>
  <input type="email" id="email" maxlength="50" placeholder="e.g. a@b.com" />
</div>
<div>
  <label for="textarea">Enter comment:</label>
  <textarea id="textarea" maxlength="400">This is a comment.</textarea>
</div>
```

Note how:

- The fields have a [`maxlength`](/en-US/docs/Web/HTML/Element/input#maxlength) attribute set, which works usefully alongside `field-sizing: content` — the size of the field will stop increasing when the character limit is reached.

> - The `<textarea>`'s `maxlength` is bigger than that of the other two fields. The result is that the `<textarea>` will grow in the inline direction until the edge of the {{cssxref("min-width")}} constraint is reached, then start to add new lines in the block direction until the `maxlength` limit is reached.

Author: Changed; I have also removed the similar text in the bullet below it as I don't think it is needed in both places.

> - The `<textarea>`'s `maxlength` is bigger than that of the other two fields. The result is that the `<textarea>` will grow in the inline direction until the edge of the {{cssxref("min-width")}} constraint is reached, then start to add new lines in the block direction until the `maxlength` limit is reached.

Reviewer 2: Agree

- The `email` input has a placeholder set. This causes the field to render big enough to show the entire placeholder. Once the field is focused and the user starts typing, the field changes size to the {{cssxref("min-width")}} set in the CSS. The `text` field, which doesn't have a placeholder, renders initially at the `min-width`.

#### CSS

> In the CSS, we set `field-sizing: content` on the three form fields, along with a {{cssxref("min-width")}} and {{cssxref("min-width")}} to constrain the input size. It is worth reiterating that, if no minimum width is set on the field, they will only be rendered as big as the text cursor.

Author: I basically used this, but I left this comment open to remind me to correct the second instance of `min-width` to `max-width` ;-)

I also changed the last sentence to:

> It is worth reiterating that, if no minimum width was set on the fields, they would be rendered only as wide as the text cursor.

We also give the `<label>`s some rudimentary styling so that they sit neatly next to the fields.

```css hidden
body {
  box-sizing: border-box;
  padding: 20px;
}

div {
  margin-bottom: 20px;
  display: flex;
}
```

```css
input,
textarea {
  field-sizing: content;
  min-width: 50px;
  max-width: 350px;
}

label {
  width: 150px;
  margin-right: 20px;
  text-align: right;
}
```

#### Result

The output looks like this. Try entering and removing text in a supporting browser to get a feel of the effects of `field-sizing: content`.

{{ EmbedLiveSample('Growing/shrinking text fields', '100%', '200') }}

### Controlling select element display

This example illustrates the effect of `field-sizing: content` on {{htmlelement("select")}} elements, both drop-down menu types and multiline listbox types.

#### HTML

> Our HTML contains two sets of `<select>` elements — one that will have `field-sizing: content` applied to it, and one that won't — so you can see the difference (the effect is a bit less obvious than on text fields). Each set contains one drop-down menu type and multiline listbox type (with the [`multiple`](/en-US/docs/Web/HTML/Element/select#multiple) attribute set).

Reviewer 2: ```suggestion
The HTML code contains two sets of `<select>` elements: one with `field-sizing: content` applied, and one without, allowing you to see the difference (though the effect may be less obvious than on text fields). Each set includes one drop-down menu type and one multiline listbox type (with the [`multiple`](/en-US/docs/Web/HTML/Element/select#multiple) attribute set).
```

```html
<div class="field-sizing">

>   <h2>With field-sizing: content</h2>

Reviewer 2: ```suggestion
  <h2>With <pre>field-sizing: content</pre></h2>
```

> The HTML contains two sets of `<select>` elements: one with `field-sizing: content` applied, and one without, allowing you to see the difference (though the effect may be less obvious than on text fields). Each set includes one drop-down menu type and one multiline listbox type (with the [`multiple`](/en-US/docs/Web/HTML/Element/select#multiple) attribute set).

Reviewer 3: add a note about why we aren't using labels.

Author: done

  <select>
    <option>Bananas</option>
    <option>Strawberries</option>
    <option selected>Apples</option>
    <option>Raspberries</option>
    <option>Pomegranate</option>
  </select>
  <select multiple>
    <option>Bananas</option>
    <option>Strawberries</option>
    <option>Apples</option>
    <option>Raspberries</option>
    <option>Pomegranate</option>
  </select>
</div>
<div>

>   <h2>Without field-sizing: content</h2>

Reviewer 2: ```suggestion
  <h2>Without <pre>field-sizing: content</pre></h2>
```

  <select>
    <option>Bananas</option>
    <option>Strawberries</option>
    <option selected>Apples</option>
    <option>Raspberries</option>
    <option>Pomegranate</option>
  </select>
  <select multiple>
    <option>Bananas</option>
    <option>Strawberries</option>
    <option>Apples</option>
    <option>Raspberries</option>
    <option>Pomegranate</option>
  </select>
</div>
```

#### CSS

> In the CSS, we set `field-sizing: content` on the first set of `<select>` elements only.

Reviewer 2: ```suggestion
In the CSS, `field-sizing: content` is set only on the first set of `<select>` elements.
```

```css hidden
body {
  box-sizing: border-box;
  display: flex;
  gap: 20px;
  font-family: sans-serif;
}

h2 {
  margin-top: 0;
  font-size: 1rem;
  text-align: center;
  flex: 1 0 100%;
}

div {
  margin-bottom: 20px;
  flex: 1;
  display: flex;
  align-items: flex-start;
  justify-content: space-around;
  flex-flow: row wrap;
}

select {
  valign: top;
}
```

```css
.field-sizing select {
  field-sizing: content;
}
```

#### Result

> The output looks like this.

Reviewer 2: ```suggestion

```

{{ EmbedLiveSample('Controlling select element display', '100%', '150') }}

> Note the effects of `field-sizing: content`:

Reviewer 2: ```suggestion
Note the following effects of `field-sizing: content`:
```

> - The drop-down menu always fits the size of the displayed option, changing size as different options are selected. Without `field-sizing: content`, the size is fixed as big as the longest option.

Reviewer 2: ```suggestion
- The drop-down menu always fits the size of the displayed option, changing size as different options are selected. Without `field-sizing: content`, the size is fixed as wide as the longest option.
```
